File: memorylayer-sdk-typescript/tests/rpg.test.ts
import { beforeEach, describe, expect, it, vi } from "vitest";

import { MemoryLayerClient } from "../src/client.js";
import { RpgNamespace } from "../src/rpg.js";

global.fetch = vi.fn();

function okJson(data: unknown) {
  return { ok: true, status: 200, json: async () => data };
}

describe("RPG namespace", () => {
  let client: MemoryLayerClient;

  beforeEach(() => {
    vi.clearAllMocks();
    client = new MemoryLayerClient({
      baseUrl: "http://localhost:61001",
      apiKey: "test-key",
      workspaceId: "ws-123",
    });
  });

  it("is available and maps camelCase sync inputs to the API contract", async () => {
    (global.fetch as any).mockResolvedValueOnce(
      okJson({ nodes_created: 1, edges_created: 1, sync_time_ms: 2 }),
    );

    expect(client.rpg).toBeInstanceOf(RpgNamespace);
    await client.rpg.sync(
      [{
        nodeId: "src/main.ts",
        nodeType: "rpg_file",
        path: "src/main.ts",
        name: "main.ts",
        parentId: "src",
      }],
      [{ sourceId: "src", targetId: "src/main.ts", relationship: "contains" }],
      { fullSync: true, sourceCommit: "abc123", contextId: "rpg-task-42" },
    );

    const [url, init] = (global.fetch as any).mock.calls[0];
    expect(url).toBe("http://localhost:61001/v1/rpg/sync");
    const body = JSON.parse(init.body);
    expect(body.nodes[0]).toMatchObject({
      node_id: "src/main.ts",
      node_type: "rpg_file",
      parent_id: "src",
    });
    expect(body.edges[0]).toMatchObject({
      source_id: "src",
      target_id: "src/main.ts",
      relationship: "contains",
      strength: 1,
    });
    expect(body.full_sync).toBe(true);
    expect(body.context_id).toBe("rpg-task-42");
  });

  it("encodes subgraph filters as query parameters", async () => {
    (global.fetch as any).mockResolvedValueOnce(
      okJson({ nodes: [], edges: [], root_id: "src", depth: 2, total_nodes: 0, total_edges: 0 }),
    );

    await client.rpg.getSubgraph({
      nodeId: "src",
      depth: 2,
      nodeTypes: ["rpg_file", "rpg_class"],
      relationshipTypes: ["contains", "imports"],
    });

    const [rawUrl] = (global.fetch as any).mock.calls[0];
    const url = new URL(rawUrl);
    expect(url.pathname).toBe("/v1/rpg/subgraph");
    expect(url.searchParams.get("node_id")).toBe("src");
    expect(url.searchParams.get("node_types")).toBe("rpg_file,rpg_class");
    expect(url.searchParams.get("relationship_types")).toBe("contains,imports");
  });

  it("serializes delete requests and propagates OBO authority", async () => {
    (global.fetch as any)
      .mockResolvedValueOnce(okJson({ deleted: 1, not_found: 0 }))
      .mockResolvedValueOnce(okJson({ workspace_id: "ws-123", has_rpg: true }));

    await client.rpg.deleteNodes(["src/main.ts"], "rpg-task-42");
    const [, deleteInit] = (global.fetch as any).mock.calls[0];
    expect(JSON.parse(deleteInit.body)).toEqual({
      node_ids: ["src/main.ts"],
      context_id: "rpg-task-42",
    });

    await client.rpg.withAuthority({
      grantId: "g-rpg",
      subject: { type: "user", id: "alice" },
    }).getStatus();
    const [, statusInit] = (global.fetch as any).mock.calls[1];
    expect(statusInit.headers["X-Aether-Grant-ID"]).toBe("g-rpg");
    expect(statusInit.headers["X-Aether-Subject-ID"]).toBe("alice");
  });
});

File: memorylayer-sdk-go/rpg_test.go
package memorylayer

import (
	"context"
	"encoding/json"
	"net/http"
	"net/url"
	"reflect"
	"testing"
)

func TestRpgSyncEncodesRequest(t *testing.T) {
	var gotMethod, gotPath, gotSubject string
	var gotBody map[string]any
	client, _ := newTestClient(t, func(w http.ResponseWriter, req *http.Request) {
		gotMethod = req.Method
		gotPath = req.URL.Path
		gotSubject = req.Header.Get("X-Aether-Subject-ID")
		if err := json.NewDecoder(req.Body).Decode(&gotBody); err != nil {
			t.Fatalf("decode body: %v", err)
		}
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"nodes_created":1,"edges_created":1,"source_commit":"abc123"}`))
	})

	if client.Rpg == nil {
		t.Fatal("NewClient did not initialize Rpg namespace")
	}
	result, err := client.Rpg.Sync(
		context.Background(),
		[]RpgNodeInput{{
			NodeID: "src/main.go", NodeType: "rpg_file", Path: "src/main.go", Name: "main.go",
		}},
		[]RpgEdgeInput{{
			SourceID: "src", TargetID: "src/main.go", Relationship: "contains", Strength: Ptr(0.0),
		}},
		RpgSyncOptions{
			FullSync:     true,
			SourceCommit: "abc123",
			ContextID:    "rpg-task-42",
			Authority: &AuthorityContext{
				GrantID: "g-rpg",
				Subject: &PrincipalRef{Type: "user", ID: "alice"},
			},
		},
	)
	if err != nil {
		t.Fatalf("Rpg.Sync: %v", err)
	}
	if result.NodesCreated != 1 || gotMethod != http.MethodPost || gotPath != "/v1/rpg/sync" {
		t.Fatalf("result=%#v method=%s path=%s", result, gotMethod, gotPath)
	}
	if gotSubject != "alice" {
		t.Errorf("subject header = %q", gotSubject)
	}
	if gotBody["context_id"] != "rpg-task-42" || gotBody["full_sync"] != true {
		t.Errorf("body = %#v", gotBody)
	}
	nodes := gotBody["nodes"].([]any)
	node := nodes[0].(map[string]any)
	if node["node_id"] != "src/main.go" || node["node_type"] != "rpg_file" {
		t.Errorf("node = %#v", node)
	}
	edges := gotBody["edges"].([]any)
	edge := edges[0].(map[string]any)
	if edge["strength"] != float64(0) {
		t.Errorf("explicit zero edge strength was not preserved: %#v", edge)
	}
}

func TestRpgGetSubgraphEncodesFilters(t *testing.T) {
	var gotQuery url.Values
	client, _ := newTestClient(t, func(w http.ResponseWriter, req *http.Request) {
		gotQuery = req.URL.Query()
		_, _ = w.Write([]byte(`{"nodes":[],"edges":[],"root_id":"src","depth":2}`))
	})

	result, err := client.Rpg.GetSubgraph(context.Background(), RpgSubgraphOptions{
		NodeID:            "src",
		Depth:             2,
		NodeTypes:         []string{"rpg_file", "rpg_class"},
		RelationshipTypes: []string{"contains", "imports"},
	})
	if err != nil {
		t.Fatalf("Rpg.GetSubgraph: %v", err)
	}
	if result.RootID == nil || *result.RootID != "src" {
		t.Fatalf("result = %#v", result)
	}
	if got := gotQuery.Get("node_types"); got != "rpg_file,rpg_class" {
		t.Errorf("node_types = %q", got)
	}
	if got := gotQuery.Get("relationship_types"); got != "contains,imports" {
		t.Errorf("relationship_types = %q", got)
	}
}

func TestRpgEnrichDefaultsAsync(t *testing.T) {
	var gotBody map[string]any
	client, _ := newTestClient(t, func(w http.ResponseWriter, req *http.Request) {
		_ = json.NewDecoder(req.Body).Decode(&gotBody)
		_, _ = w.Write([]byte(`{"workspace_id":"ws_default","context_id":"rpg","async_mode":true}`))
	})

	result, err := client.Rpg.Enrich(context.Background(), RpgEnrichOptions{})
	if err != nil {
		t.Fatalf("Rpg.Enrich: %v", err)
	}
	if !result.AsyncMode || !reflect.DeepEqual(gotBody, map[string]any{
		"async": true, "context_id": "rpg", "phases": nil,
	}) {
		t.Fatalf("result=%#v body=%#v", result, gotBody)
	}
}
